### ISSUE TITLE
Slicing sparse tensor

### ISSUE THREAD

**Issue by eriophora** [tensorflow/tensorflow#1588]
There does not appear to be a way to extract a (sparse or dense) slice from a sparse tensor in the same manner as there is from a dense tensor. Will this be a feature in the future, or is it already present and I'm just missing something? Perhaps it should be added to the planned improvements for indexing.

**Comment by eriophora** [comment]
To expand on this, I'd like to be able to construct (ultimately dense, but this is irrelevant) subtensor from a  much larger, sparse tensor that's far too large to fit into memory. In numpy, this is something like having:

```
x = np.array(range(16)).reshape(4, 4)
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])
```

And wanting to extract the subtensor formed by the 2nd (i.e., index 1) and 4th row with the 2nd and 3rd column:

```
x[[[1],[3]],[1,2]]
array([[ 5,  6],
       [13, 14]])
```

Ultimately, the batches of the network I'm training will rely on such fixed-size subtensors randomly constructed from the larger tensor. However, at present the only way of doing this currently appears to be manually feeding the subtensors in from Scipy.

**Comment by ebrevdo** [comment]
We don't currently have any plans to implement a slice op on SparseTensor.  My suggestion would be to first implement a sparse->sparse slice, and then use sparse_to_dense to densify it.

The closest thing we have is a sparse split; the meat of which is implemented here:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/sparse/sparse_tensor.h#L437

If you're interested in implementing a SparseSlice, contributions are welcome.  The signature should probably be:

SparseTensor SparseTensor::Slice(const SparseTensor& input_tensor, const gtl::ArraySlice<int64>& start, const gtl::ArraySlice<int64>& size)

**Comment by mrry** [comment]
@concretevitamin: Is slice one of the ops for which you're implementing SparseTensor support?

**Comment by ebrevdo** [comment]
mdan is implementing SparseTensor slice this week.  I think there's some
code already done.

**Comment by concretevitamin** [comment]
I am not actively working on this op.  Happy to review any PRs ;)

**Comment by joergfranke** [comment]
Is someone working on this op?

**Comment by ebrevdo** [comment]
No, you are welcome to give it a shot.

**Comment by soloice** [comment]
@ebrevdo 
I've read a related [comment](https://github.com/tensorflow/tensorflow/issues/342#issuecomment-160354041) of yours and it helped me a lot.

Is it possible to extract sparse ids and train sparse weights on the fly?

My application is finetuning sparse embedding: Given a sequence of word ids, firstly use a cookbook to look up sparse indices and weights of each word, then use `tf.nn.embedding_lookup_sparse` to get their embeddings. Since I want to finetune the sparse weights,  directly feed them to placeholders are not adequate.

To be more specific:
1. How can I extract certain rows (maybe discrete and even have duplicate rows) out of a `SparseTensor`?
2. If the first one is possible, can I perform training on sparse weights? (Words may duplicate in a sentence, so sparse weights may duplicate in a sentence, so the variables need to be shared to make sure gradient descent algorithms work properly. Simply pass pre-calculated weights to a placeholder does not work for this purpose)

Currently is it possible to do this in TensorFlow? Or I need some work around (e.g. use sparse dense matrix multiplication or something else) to do this?

**Comment by yongtang** [comment]
Added a PR #9540 to support slice of sparse tensors. Please take a look.

**Comment by pstjohn** [comment]
What would it take to implement a `SparseTensor.__getitem__` method?
There's no `tf.sparse_strided_slice` function which `Tensor.__getitem__` seems to use.
